LFS171x/fabric-material/chaincode/tuna-app/tuna-chaincode.go

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: Apache-2.0
+
+/*
+ * Sample Chaincode based on Demonstrated Scenario
+ */
+
+package main
+
+/* Imports
+ * 4 utility libraries for formatting, handling bytes, reading and writing JSON, and string manipulation
+ * 2 specific Hyperledger Fabric specific libraries for Smart Contracts
+ */
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/hyperledger/fabric/core/chaincode/shim"
+	sc "github.com/hyperledger/fabric/protos/peer"
+)
+
+// Define the Smart Contract structure
+type SmartContract struct {
+}
+
+// Define Tuna structure, with 4 properties.  Structure tags are used by encoding/json library
+type Tuna struct {
+	Vessel string `json:"vessel"`
+	Timestamp string `json:"timestamp"`
+	Location  string `json:"location"`
+	Holder  string `json:"holder"`
+}
+
+/*
+ * The Init method is called when the Smart Contract "tuna-chaincode" is instantiated by the blockchain network
+ * Best practice is to have any Ledger initialization in separate function -- see initLedger()
+ */
+func (s *SmartContract) Init(APIstub shim.ChaincodeStubInterface) sc.Response {
+	return shim.Success(nil)
+}
+
+/*
+ * The Invoke method is called when an application requests to run the Smart Contract "tuna-chaincode"
+ * The application also specifies the particular smart contract function to call with arguments
+ */
+func (s *SmartContract) Invoke(APIstub shim.ChaincodeStubInterface) sc.Response {
+
+	// Retrieve the requested Smart Contract function and arguments
+	function, args := APIstub.GetFunctionAndParameters()
+	// Route to the appropriate handler function to interact with the ledger appropriately
+	if function == "queryTuna" {
+		return s.queryTuna(APIstub, args)
+	} else if function == "initLedger" {
+		return s.initLedger(APIstub)
+	} else if function == "recordTuna" {
+		return s.recordTuna(APIstub, args)
+	} else if function == "queryAllTuna" {
+		return s.queryAllTuna(APIstub)
+	} else if function == "changeTunaHolder" {
+		return s.changeTunaHolder(APIstub, args)
+	}
+
+	return shim.Error("Invalid Smart Contract function name.")
+}
+
+func (s *SmartContract) queryTuna(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
+
+	if len(args) != 1 {
+		return shim.Error("Incorrect number of arguments. Expecting 1")
+	}
+
+	tunaAsBytes, _ := APIstub.GetState(args[0])
+	return shim.Success(tunaAsBytes)
+}
+
+/*
+ * The Init method is called when the Smart Contract "tuna-chaincode" is instantiated by the Fabric network
```

Comment:
This comment should be updated to reflect what initLedger does.

Reply:
done